crates/stackable-operator/src/validation.rs

```diff
@@ -26,121 +27,202 @@ const RFC_1123_SUBDOMAIN_MAX_LENGTH: usize = 253;
+//
+// Hence, restrict them to a reasonable subset. The convention is to use upper-case
+// DNS hostnames, so allow all characters used there.
+const KERBEROS_REALM_NAME_FMT: &str = "[-.a-zA-Z0-9]+";
+const KERBEROS_REALM_NAME_ERROR_MSG: &str =
+    "Kerberos realm name must only contain alphanumeric characters, '-', and '.'";
+
 // Lazily initialized regular expressions
-static RFC_1123_SUBDOMAIN_REGEX: LazyLock<Regex> = LazyLock::new(|| {
+pub(crate) static RFC_1123_SUBDOMAIN_REGEX: LazyLock<Regex> = LazyLock::new(|| {
     Regex::new(&format!("^{RFC_1123_SUBDOMAIN_FMT}$"))
         .expect("failed to compile RFC 1123 subdomain regex")
 });
 
+static RFC_1123_LABEL_REGEX: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(&format!("^{RFC_1123_LABEL_FMT}$")).expect("failed to compile RFC 1123 label regex")
+});
+
 static RFC_1035_LABEL_REGEX: LazyLock<Regex> = LazyLock::new(|| {
     Regex::new(&format!("^{RFC_1035_LABEL_FMT}$")).expect("failed to compile RFC 1035 label regex")
 });
 
-/// Returns a formatted error message for maximum length violations.
-fn max_len_error(length: usize) -> String {
-    format!("must be no more than {length} characters")
-}
+pub(crate) static KERBEROS_REALM_NAME_REGEX: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(&format!("^{KERBEROS_REALM_NAME_FMT}$"))
+        .expect("failed to compile Kerberos realm name regex")
+});
 
-/// Returns a formatted error message for regex violations.
-///
-/// # Arguments
-///
-/// * `msg` - this is the main error message to return
-/// * `fmt` - this is the regular expression that did not match the input
-/// * `examples` - are optional well, formed examples that would match the regex
-fn regex_error(msg: &str, fmt: &str, examples: &[&str]) -> String {
-    if examples.is_empty() {
-        return format!("{msg} (regex used for validation is '{fmt}')");
-    }
+type Result<T = (), E = Errors> = std::result::Result<T, E>;
+
+/// A collection of errors discovered during validation.
+#[derive(Debug)]
+pub struct Errors(Vec<Error>);
 
-    let mut msg = msg.to_string();
-    msg.push_str(" (e.g. ");
-    for (i, example) in examples.iter().enumerate() {
-        if i > 0 {
-            msg.push_str(" or ");
+impl Display for Errors {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for (i, error) in self.0.iter().enumerate() {
+            let prefix = match i {
+                0 => "",
+                _ => ", ",
+            };
+            write!(f, "{prefix}{error}")?;
         }
-        msg.push('\'');
-        msg.push_str(example);
-        msg.push_str("', ");
+        Ok(())
     }
+}
+impl std::error::Error for Errors {}
+
+/// A single validation error.
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(transparent)]
+    Regex { source: RegexError },
+    #[snafu(display("input is {length} bytes long but must be no more than {max_length}"))]
+    TooLong { length: usize, max_length: usize },
```

Comment:
style: Please add newlines between variants. See [here](https://docs.stackable.tech/home/stable/contributor/code-style-guide#_formatting_of_struct_fields_and_enum_variants).

```suggestion
    #[snafu(transparent)]
    Regex { source: RegexError },

    #[snafu(display("input is {length} bytes long but must be no more than {max_length}"))]
    TooLong { length: usize, max_length: usize },
```